### PR TITLE
Add microphone selection UI and error handling in audio recording

### PR DESCRIPTION
**Pull request by ItsSimko**
## Summary by Sourcery

Add a microphone selection UI to enable users to choose their preferred microphone for audio recording. Implement error handling for audio stream opening failures, providing user feedback when issues arise with microphone settings. Integrate the microphone selection into the settings window to allow users to save their preferences.

New Features:
- Introduce a microphone selection UI component to allow users to choose their preferred microphone for audio recording.

Bug Fixes:
- Add error handling for audio stream opening failures, providing user feedback when microphone settings are incorrect.

Enhancements:
- Integrate the microphone selection feature into the settings window, allowing users to save their microphone preferences.

### CODE REVIEW

**Comment by ItsSimko** [comment]
@sourcery-ai review

File: src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+    ----------
+    PYAUDIO : pyaudio.PyAudio
+        A PyAudio instance used to interact with audio devices.
+    SELECTED_MICROPHONE_INDEX : int
+        The index of the currently selected microphone.
+    SELECTED_MICROPHONE_NAME : str
+        The name of the currently selected microphone.
+    """
+
+    PYAUDIO = pyaudio.PyAudio()
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py:19]: i think this is OK so we can access them. We really only have one dropdown option for microphone anyways.

File: src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    SELECTED_MICROPHONE_INDEX = 1
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py:20]: Commit 6facd1eb6d7c4c9b48724689244b1121a4001025

File: src/FreeScribe.client/UI/Widgets/AudioMeter.py [hunk 3]
```diff
@@ -196,9 +197,10 @@ def toggle_monitoring(self):
             self.running = True
             self.stream = self.p.open(
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/Widgets/AudioMeter.py:198]: COmmit f4dbf523b37c8dda895144b5f3679f7c421c4b38

File: src/FreeScribe.client/client.py [hunk 4]
```diff
@@ -157,7 +158,20 @@ def toggle_pause():
...
+
+    try:
+        stream = p.open(
+            format=FORMAT, 
+            channels=1, 
+            rate=RATE, 
+            input=True,
+            frames_per_buffer=CHUNK, 
+            input_device_index=int(MicrophoneSelector.SELECTED_MICROPHONE_INDEX))
+    except Exception as e:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/client.py]: Commited

File: src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
+import tkinter as tk
+from tkinter import ttk
+import pyaudio
+
+class MicrophoneSelector:
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py:5]: Convert class variables to instance variables:
 Refreenced in earlier issue. I think its ok for class variables for these

Rest here
550581568c1feca4de86a86d72329e06a263061c

**Comment by ItsSimko** [comment]
@sourcery-ai review

File: src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
```
[the remark this answers is not in the document]
Comment (ItsSimko) [reply src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py]: f971d9c9c9a07ad4e918e7db7e85206d2b5de072

File: src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yingbull) [reply src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py:37]: It is still OK to suggest refactoring, sourcery, but in this case this was just a better fit.

File: src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            Application settings including editable settings.
+
+        Returns
+        -------
+        str
+            The name of the currently selected microphone.
+        """
+        if "Current Mic" in app_settings.editable_settings:
+            MicrophoneSelector.SELECTED_MICROPHONE_NAME = app_settings.editable_settings["Current Mic"]
+            for i in range(MicrophoneSelector.PYAUDIO.get_device_count()):
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py]: Generally in favor of try/catch blocks.  Sourcery, open an issue ticket to track looking at this a bit closer.

File: src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (yingbull) [reply src/FreeScribe.client/UI/Widgets/MicrophoneSelector.py]: Good change.

**Review by yingbull — APPROVED** [review]

**Pull request merged**